### ISSUE TITLE
`TypeError` in jax device docstring example

### ISSUE THREAD

**Issue by albi3ro** [PennyLaneAI/pennylane#1236]
Upon executing the example code in [the jax device docstring](https://github.com/PennyLaneAI/pennylane/blob/8e55e3dafa16e3f68e3717459e1ce176db9b602c/pennylane/devices/default_qubit_jax.py#L83), 

```python
dev = qml.device("default.qubit.jax", wires=1)

@jax.jit
@qml.qnode(dev, interface="jax", diff_method="backprop")
def circuit():
    qml.Hadamard(0)
    return qml.sample(qml.PauliZ(wires=0))

a = circuit()
b = circuit() # Bad! b will be the exact same samples as a.
```

I am getting ``TypeError: unsupported operand type(s) for *=: 'int' and 'NoneType'``. 

Any ideas? @Thenerdstation

**Comment by albi3ro** [comment]
Still occurs when I remove the jit decorator.

The error output is:

```
WARNING:absl:No GPU/TPU found, falling back to CPU. (Set TF_CPP_MIN_LOG_LEVEL=0 and rerun for more info.)
---------------------------------------------------------------------------
FilteredStackTrace                        Traceback (most recent call last)
<ipython-input-3-9beb53ba2a5f> in <module>
      8 
----> 9 a = circuit()
     10 b = circuit() # Bad! b will be the exact same samples as a.

/mnt/c/Users/chris/Prog/xanadu/pennylane/pennylane/qnode.py in __call__(self, *args, **kwargs)
    554         # execute the tape
--> 555         res = self.qtape.execute(device=self.device)
    556 

/mnt/c/Users/chris/Prog/xanadu/pennylane/pennylane/tape/tape.py in execute(self, device, params)
   1233 
-> 1234         return self._execute(params, device=device)
   1235 

/mnt/c/Users/chris/Prog/xanadu/pennylane/pennylane/tape/tape.py in execute_device(self, params, device)
   1264         if isinstance(device, qml.QubitDevice):
-> 1265             res = device.execute(self)
   1266         else:

/mnt/c/Users/chris/Prog/xanadu/pennylane/pennylane/_qubit_device.py in execute(self, circuit, **kwargs)
    187         if self.shots is not None or circuit.is_sampled:
--> 188             self._samples = self.generate_samples()
    189 

/mnt/c/Users/chris/Prog/xanadu/pennylane/pennylane/_qubit_device.py in generate_samples(self)
    452 
--> 453         samples = self.sample_basis_states(number_of_states, rotated_prob)
    454         return QubitDevice.states_to_binary(samples, self.num_wires)

/mnt/c/Users/chris/Prog/xanadu/pennylane/pennylane/devices/default_qubit_jax.py in sample_basis_states(self, number_of_states, state_probability)
    248             key = self._prng_key
--> 249         return jax.random.choice(key, number_of_states, shape=(self.shots,), p=state_probability)
    250 

~/miniconda3/envs/xanadu/lib/python3.8/site-packages/jax/_src/random.py in choice(key, a, shape, replace, p)
    608   n_inputs = int(a) if np.ndim(a) == 0 else len(a)  # type: ignore[arg-type]
--> 609   n_draws = prod(shape)
    610   if n_draws == 0:

FilteredStackTrace: TypeError: unsupported operand type(s) for *=: 'int' and 'NoneType'

The stack trace above excludes JAX-internal frames.
The following is the original exception that occurred, unmodified.

--------------------

The above exception was the direct cause of the following exception:

TypeError                                 Traceback (most recent call last)
<ipython-input-3-9beb53ba2a5f> in <module>
      7     return qml.sample(qml.PauliZ(wires=0))
      8 
----> 9 a = circuit()
     10 b = circuit() # Bad! b will be the exact same samples as a.

~/miniconda3/envs/xanadu/lib/python3.8/site-packages/jax/_src/traceback_util.py in reraise_with_filtered_traceback(*args, **kwargs)
    137   def reraise_with_filtered_traceback(*args, **kwargs):
    138     try:
--> 139       return fun(*args, **kwargs)
    140     except Exception as e:
    141       if not is_under_reraiser(e):

~/miniconda3/envs/xanadu/lib/python3.8/site-packages/jax/api.py in cache_miss(*args, **kwargs)
    330       _check_arg(arg)
    331     flat_fun, out_tree = flatten_fun(f, in_tree)
--> 332     out_flat = xla.xla_call(
    333         flat_fun,
    334         *args_flat,

~/miniconda3/envs/xanadu/lib/python3.8/site-packages/jax/core.py in bind(self, fun, *args, **params)
   1400 
   1401   def bind(self, fun, *args, **params):
-> 1402     return call_bind(self, fun, *args, **params)
   1403 
   1404   def process(self, trace, fun, tracers, params):

~/miniconda3/envs/xanadu/lib/python3.8/site-packages/jax/core.py in call_bind(primitive, fun, *args, **params)
   1391   tracers = map(top_trace.full_raise, args)
   1392   with maybe_new_sublevel(top_trace):
-> 1393     outs = primitive.process(top_trace, fun, tracers, params)
   1394   return map(full_lower, apply_todos(env_trace_todo(), outs))
   1395 

~/miniconda3/envs/xanadu/lib/python3.8/site-packages/jax/core.py in process(self, trace, fun, tracers, params)
   1403 
   1404   def process(self, trace, fun, tracers, params):
-> 1405     return trace.process_call(self, fun, tracers, params)
   1406 
   1407   def post_process(self, trace, out_tracers, params):

~/miniconda3/envs/xanadu/lib/python3.8/site-packages/jax/core.py in process_call(self, primitive, f, tracers, params)
    598 
    599   def process_call(self, primitive, f, tracers, params):
--> 600     return primitive.impl(f, *tracers, **params)
    601   process_map = process_call
    602 

~/miniconda3/envs/xanadu/lib/python3.8/site-packages/jax/interpreters/xla.py in _xla_call_impl(fun, device, backend, name, donated_invars, *args)
    574 
    575 def _xla_call_impl(fun: lu.WrappedFun, *args, device, backend, name, donated_invars):
--> 576   compiled_fun = _xla_callable(fun, device, backend, name, donated_invars,
    577                                *unsafe_map(arg_spec, args))
    578   try:

~/miniconda3/envs/xanadu/lib/python3.8/site-packages/jax/linear_util.py in memoized_fun(fun, *args)
    258       fun.populate_stores(stores)
    259     else:
--> 260       ans = call(fun, *args)
    261       cache[key] = (ans, fun.stores)
    262 

~/miniconda3/envs/xanadu/lib/python3.8/site-packages/jax/interpreters/xla.py in _xla_callable(fun, device, backend, name, donated_invars, *arg_specs)
    650 
    651   abstract_args, arg_devices = unzip2(arg_specs)
--> 652   jaxpr, out_avals, consts = pe.trace_to_jaxpr_final(fun, abstract_args, transform_name="jit")
    653   if any(isinstance(c, core.Tracer) for c in consts):
    654     raise core.UnexpectedTracerError("Encountered an unexpected tracer.")

~/miniconda3/envs/xanadu/lib/python3.8/site-packages/jax/interpreters/partial_eval.py in trace_to_jaxpr_final(fun, in_avals, transform_name)
   1207     main.source_info = fun_sourceinfo(fun.f, transform_name)  # type: ignore
   1208     main.jaxpr_stack = ()  # type: ignore
-> 1209     jaxpr, out_avals, consts = trace_to_subjaxpr_dynamic(fun, main, in_avals)
   1210     del fun, main
   1211   return jaxpr, out_avals, consts

~/miniconda3/envs/xanadu/lib/python3.8/site-packages/jax/interpreters/partial_eval.py in trace_to_subjaxpr_dynamic(fun, main, in_avals)
   1186     trace = DynamicJaxprTrace(main, core.cur_sublevel())
   1187     in_tracers = map(trace.new_arg, in_avals)
-> 1188     ans = fun.call_wrapped(*in_tracers)
   1189     out_tracers = map(trace.full_raise, ans)
   1190     jaxpr, out_avals, consts = frame.to_jaxpr(in_tracers, out_tracers)

~/miniconda3/envs/xanadu/lib/python3.8/site-packages/jax/linear_util.py in call_wrapped(self, *args, **kwargs)
    164 
    165     try:
--> 166       ans = self.f(*args, **dict(self.params, **kwargs))
    167     except:
    168       # Some transformations yield from inside context managers, so we have to

/mnt/c/Users/chris/Prog/xanadu/pennylane/pennylane/qnode.py in __call__(self, *args, **kwargs)
    553 
    554         # execute the tape
--> 555         res = self.qtape.execute(device=self.device)
    556 
    557         if original_shots is not None:

/mnt/c/Users/chris/Prog/xanadu/pennylane/pennylane/tape/tape.py in execute(self, device, params)
   1232             params = self.get_parameters()
   1233 
-> 1234         return self._execute(params, device=device)
   1235 
   1236     def execute_device(self, params, device):

/mnt/c/Users/chris/Prog/xanadu/pennylane/pennylane/tape/tape.py in execute_device(self, params, device)
   1263 
   1264         if isinstance(device, qml.QubitDevice):
-> 1265             res = device.execute(self)
   1266         else:
   1267             res = device.execute(self.operations, self.observables, {})

/mnt/c/Users/chris/Prog/xanadu/pennylane/pennylane/_qubit_device.py in execute(self, circuit, **kwargs)
    186         # generate computational basis samples
    187         if self.shots is not None or circuit.is_sampled:
--> 188             self._samples = self.generate_samples()
    189 
    190         # compute the required statistics

/mnt/c/Users/chris/Prog/xanadu/pennylane/pennylane/_qubit_device.py in generate_samples(self)
    451         rotated_prob = self.analytic_probability()
    452 
--> 453         samples = self.sample_basis_states(number_of_states, rotated_prob)
    454         return QubitDevice.states_to_binary(samples, self.num_wires)
    455 

/mnt/c/Users/chris/Prog/xanadu/pennylane/pennylane/devices/default_qubit_jax.py in sample_basis_states(self, number_of_states, state_probability)
    247         else:
    248             key = self._prng_key
--> 249         return jax.random.choice(key, number_of_states, shape=(self.shots,), p=state_probability)
    250 
    251     @staticmethod

~/miniconda3/envs/xanadu/lib/python3.8/site-packages/jax/_src/random.py in choice(key, a, shape, replace, p)
    607     a = _asarray(a)
    608   n_inputs = int(a) if np.ndim(a) == 0 else len(a)  # type: ignore[arg-type]
--> 609   n_draws = prod(shape)
    610   if n_draws == 0:
    611     return jnp.zeros(shape, dtype=lax.dtype(a))

~/miniconda3/envs/xanadu/lib/python3.8/site-packages/jax/_src/util.py in prod(xs)
    206   out = 1
    207   for x in xs:
--> 208     out *= x
    209   return out
    210 

TypeError: unsupported operand type(s) for *=: 'int' and 'NoneType'
```

**Comment by albi3ro** [comment]
With `qml.sample`, the number of shots must be specified.  The docstring is getting updated with a finite number of shots for the device, and the jax device should throw a better error in this situation.

**Comment by albi3ro** [comment]
Fixed by PR #1238
